src/Components/Selection.re

```diff
@@ -0,0 +1,75 @@
+[@deriving show({with_path: false})]
+type t = {
+  anchor: int,
+  focus: int,
+};
+
+type directionVariant =
+  | Left(int)
+  | Right(int);
+
+let initial: t = {anchor: 0, focus: 0};
+
+let withinLength = (text: string, position: int): int => {
+  let withingStart = max(position, 0);
+  let withingBoth = min(withingStart, String.length(text));
+
+  withingBoth;
+};
+
+let create = (~text: string, ~anchor: int, ~focus: int): t => {
+  let safeAnchor = withinLength(text, anchor);
+  let safeFocus = withinLength(text, focus);
+
+  {anchor: safeAnchor, focus: safeFocus};
+};
+
+let anchor = (selection: t): int => {
+  selection.anchor;
+};
+
+let focus = (selection: t): int => {
+  selection.focus;
+};
+
+let range = (selection: t): int => {
+  abs(selection.focus - selection.anchor);
+};
+
+let rangeStart = (selection: t): int => {
+  min(selection.focus, selection.anchor);
+};
+
+let rangeEnd = (selection: t): int => {
+  max(selection.focus, selection.anchor);
+};
+
+let isCollapsed = (selection: t): bool => {
+  selection.anchor == selection.focus;
+};
+
+let collapse = (~text: string, offset: int): t => {
+  create(~text, ~anchor=offset, ~focus=offset);
+};
+
+let collapseRelative =
+    (~text: string, ~selection: t, direction: directionVariant): t => {
+  switch (direction) {
+  | Left(offset) => collapse(~text, rangeStart(selection) - abs(offset))
+  | Right(offset) => collapse(~text, rangeEnd(selection) + abs(offset))
+  };
+};
```

Comment:
This seems more like creating a new selection relative to the old one, than collapsing it. Or alternatively collapse an move, but then I think it should remain as two operations.

Reply:
I think concept of `collapseRelative` have a lot of advantages. 
It take `selection` and just collapse thing relatively in direction.

renaming it to `collapseAndMove` will be less representative from my point.
And introducing `move`  will require to add some logic in place where we will use it to calculate offset (but this will turn out to just using collapsed method)

So I'd better leave this thing here. 
But If you would insist I can change it :)

Reply:
> And introducing move will require to add some logic in place where we will use it to calculate offset (but this will turn out to just using collapsed method)

No, `move` would just move the selection. Increasing or decreasing both `anchor` and `offset` the same amount. No collapsing, that's a different operation.

> It take selection and just collapse thing relatively in direction.

That's just `collapse` in my mind. There is no "absolute" collapse, it's always relative, so I think the right approach is to just name it `collapse` and then pass in a variant that specify where it should collapse to. Either left, right, anchor or focus.

From a use POV, I'd prefer an API that looked something like this instead of `collapseRelative`:

```
selection |> collapse(`Left) |> move(1)
```

or 

```
selection |> collapseTo(`Anchor) |> move(`Right, 1)
```

Do you see how this kind of compositional API can be both simpler _and_ more flexible?

Reply:
I've found better solution. I just removed those `relative` things. They kinda redundant.

Reply:
I'd prefer the simplicity and flexibility of a compositional API, because code tends to evolve, but yeah, that solution works fine :+1: